File: framework/test2/struct/query2d/Query2d.scala
import soda.scala.unittest._

import scala.reflect.runtime.universe.typeOf

import soda.scala.leetcode._

import collection.mutable
import collection.mutable.{ArrayBuffer, ArrayDeque, PriorityQueue}
import util.control.Breaks.{break, breakable}

class NumMatrix(matrix: Array[Array[Int]]) {

  private var bit: Array[Array[Int]] = Array.empty
  private var rows: Int = 0
  private var cols: Int = 0

  val _init = () => {
    rows = matrix.length + 1
    cols = matrix(0).length + 1
    bit = Array.ofDim[Int](rows, cols)
    for (i <- 0 until rows - 1) {
      for (j <- 0 until cols - 1) {
        val value = matrix(i)(j)
        matrix(i)(j) = 0
        update(i, j, value)
      }
    }
  }
  _init()

  def update(row: Int, col: Int, `val`: Int): Unit = {
    val diff = `val` - matrix(row)(col)
    matrix(row)(col) = `val`
    var i = row + 1
    while (i < rows) {
      var j = col + 1
      while (j < cols) {
        bit(i)(j) += diff
        j += (j & -j)
      }
      i += (i & -i)
    }
  }

  def sumRegion(row1: Int, col1: Int, row2: Int, col2: Int): Int = {
    query(row1,col1) - query(row1,col2+1) - query(row2+1,col1) + query(row2+1,col2+1)
  }

  def query(r: Int, c: Int): Int = {
    var res = 0
    var i = r
    while (i > 0) {
      var j = c
      while (j > 0) {
        res += bit(i)(j)
        j -= (j & -j)
      }
      i -= (i & -i)
    }
    res
  }

}

class Query2d {
  def get(): () => Unit = {
    () => {
      // val work = GenericTestWork.create1(Solution.eq)
      val work = GenericTestWork.forStruct(typeOf[NumMatrix])
      // val work = TestWork.forObject(typeOf[Solution], "METHOD")
      // val work = TestWork.forInstance(typeOf[Solution], "METHOD")
      // val work = TestWork.createN((...)=>R)
      // val work = TestWork.forStruct(typeOf[STRUCT])
      // work.setValidator((R, R) => Boolean)
      work.compareSerial = true
      work.run()
    }
  }
}

object Query2d extends App {
  new Query2d().get().apply()
}

File: framework/test2/struct/query2d/Query2d.java
import soda.unittest.*;

import java.util.*;
import java.util.function.Supplier;
import java.util.stream.*;

import soda.leetcode.*;
import soda.unittest.Validators;
import soda.unittest.TestWork;


class NumMatrix {

    private int[][] matrix;

    private int[][] bit;

    private int rows, cols;

    public NumMatrix(int[][] matrix) {
        this.matrix = matrix;
        rows = matrix.length + 1;
        cols = matrix[0].length + 1;
        bit = new int[rows][cols];
        for (int i = 0; i < rows - 1; ++i) {
            for (int j = 0; j < cols - 1; ++j) {
                int val = matrix[i][j];
                matrix[i][j] = 0;
                update(i, j, val);
            }
        }
    }
    
    public void update(int row, int col, int val) {
        int diff = val - this.matrix[row][col];
        this.matrix[row][col] = val;
        for (int i = row+1; i < this.rows; i += (i & -i)) {
            for (int j = col+1; j < this.cols; j += (j & -j)) {
                this.bit[i][j] += diff;
            }
        }
    }
    
    public int sumRegion(int row1, int col1, int row2, int col2) {
        return query(row1, col1) - query(row1, col2+1) - query(row2+1, col1) + query(row2+1, col2+1);
    }

    private int query(int r, int c) {
        int res = 0;
        for (int i = r; i > 0; i -= (i & -i)) {
            for (int j = c; j > 0; j -= (j & -j)) {
                res += bit[i][j];
            }
        }
        return res;
    }
}

public class Query2d implements Supplier<TestWork> {

    @Override
    public TestWork get() {
        var work = TestWork.forStruct(NumMatrix.class);
        // work.setValidator((e, r) -> {...});
        work.setCompareSerial(true);
        return work;
    }

    public static void main(String[] args) throws Exception {
        new Query2d().get().run();
    }

}
